만국박람회 [STEP 2] Maxhyunm, Whales

WhalesJin opened the pull request:
안녕하세요, Steven(@stevenkim18)🦦!!
만국박람회 Step2 PR로 인사드리는 👍🏻최고래조-Max🤖 & Whales🐋입니다!!
저희는 이번 스텝을 진행하면서 둘 다 처음으로 **스토리보드 없이 코드로만 UI를 구현**하는 도전을 해보았는데요, 덕분에 굉장히 알차고 알찬, 일주일 같은 이틀을 보냈습니다!!!
사이사이 많이 막히기도 했지만 그 때마다 또 해결도 하면서 둘이서 굉장히 즐겁고 뿌듯한 프로젝트를 진행하였고 그로 인해 많은 공부가 되었습니다.

그럼에도 불구하고 아직은 부족한 부분이 많아요.
보시면서 아쉬운 부분, 잘못된 부분 혹은 더 공부해야 할 부분에 대한 조언을 주신다면 감사한 마음으로 더 열심히 해보려합니다!!

이번 리뷰도 잘 부탁드립니다 Steven 🙇🏻‍♀️


## 🤹🏻 고민했던 점
>**이모지로 표시한 부분**
고민한 점 - 🤔
문제점 - 🚨
해결 - 💡
### 🔹 UI 구현 - Storyboard vs Code
🤔 스토리보드는 이미 많은 부분을 자동으로 해주고 있어서 참 편리한 친구라고 생각하지만 스토리보드로만 할 수 있는 부분에 대해서는 한계가 있기에 코드로 구현하는 부분들을 더 활용하고 싶어서 공부를 하고 적용해 보았습니다.
- 스토리보드와 코드를 비교했을 때 코드의 장점
    - 나중에 소스로 볼 때
        - 스토리보드 : XML이라 보기 힘들다
        - 코드 : 코드로 적은 거라 더 깔끔하고 이해하기 좋다
    - 빌드
        - 스토리보드 -> 코드 -> 시뮬레이터
        - 코드 -> 시뮬레이터 (빌드 속도가 좀 더 빠르다)
- 코드베이스로 구현한 이유
    - 커스텀으로 `TableViewCell`과 `View`들을 만들어 적용해보기 위해(`View` 파일 분리)
    - 기존에 스토리보드 형태로만 작업을 해 보았기 때문에 새로운 학습을 위해 시도
- 스토리보드 삭제 진행 과정
    - 삭제할 항목
        - Main.storyboard
        - info.plist -> ... -> Storyboard Name
        - info.plist -> Main storyboard file base name

🚨 필요한 항목을 삭제하고 나니 앱이 가장 처음 보여줄 `ViewController`를 찾지 못해 빌드 오류가 발생했습니다.

💡 `SceneDelegate`의 `scene 메서드` 내부에 다음 코드를 삽입하여 해결하였습니다.(`NavigationController`를 위한 추가 구현 포함)
```swift
guard let windowScene = (scene as? UIWindowScene) else { return }
window = UIWindow(windowScene: windowScene)
let mainViewController = ViewController()
let navigationController = UINavigationController (rootViewController: mainViewController)
window?.rootViewController = navigationController
window?.makeKeyAndVisible()
```

### 🔹 Navigation vs Modality
- 화면의 흐름이 `계층구조`로 이루어져 있다고 판단했습니다.
    `메인. 한국의 출품작 보기 -> 한국의 출품작. 작품 선택 -> 해당 작품의 상세화면`
- 요구사항 화면에 네비게이션 바 버튼을 통해 화면을 이동하는 부분이 포함되어 있기에 최종적으로 네비게이션 방식을 선택하였습니다.


### 🔹 레이아웃 관리
- **`StackView` 내부 요소가 화면 영역보다 커지는 경우**

    🚨 `StackView`의 `leftAnchor`, `rightAnchor`를 맞추었음에도 불구하고 `StackView` 내부 요소가 커지면 상위 요소들이 함께 늘어나는 것을 확인했습니다.
    
    💡 `StackView`의 `width`를 지정하여 해결하였습니다.
    ```swift
    mainStackView.widthAnchor.constraint(equalTo: mainScrollView.widthAnchor, constant: -40)
    ```
- **`StackView` 내부 요소가 잘리거나 사라지는 경우**

    🚨 `horizontal`로 지정된 `StackView`에서 하나의 요소가 커지면 다른 요소가 잘리거나 사라지는 것을 확인했습니다.
    
    💡 `setContentCompressionResistancePriority` 설정으로 텍스트버튼의 resistance 우선순위를 높였습니다.
    
    💡 태극기의 높이를 버튼 높이와 맞추고, 태극기의 가로 길이는 높이와 비율을 계산하여 맞추어 해결하였습니다.


### 🔹 ViewController 기능 분리
🤔 요소를 구현할 때마다 관련 인스턴스를 생성하여 추가하고 제약조건을 넣다 보니 ViewController가 너무 무거워지는 것을 느꼈습니다. 때문에 아래 방법들로 ViewController의 정리 및 기능 분리를 시도했습니다.
- View 파일 분리
    - 내부 요소로 추가되는 중첩 `StackView`를 커스텀 `StackView`로 분리했습니다.
        - `MainStackView`
        - `ExpositionInformationStackView`
        - `ChangeViewButtonStackView`
    - 데이터 형태에 맞춰 `TableViewCell`을 유연하게 관리하기 위해 커스텀 Cell을 만들어 분리했습니다.
        - `ExpositionItemTableViewCell`
- DecodingManager 분리
    - 데이터를 불러오는 작업만을 `DecodingManager`로 분리하여 어디에서든 접근 가능하도록 만들었습니다.
- 제약조건 설정 메서드 분리
    - 제약조건 관련 내용을 하나의 메서드에 담아 `ExpositionViewController`의 `extension`으로 분리해 추가했습니다.
- Namespace 생성
    - 매직리터럴로 작성된 파일명이나 타이틀명 등을 추후 유지보수를 위하여 분리했습니다.
        - `ViewControllerTitleNamespace`
        - `AssetNamespace`
        - `LabelTextNamespace`
        - `IdentifierNamespace`

### 🔹 ViewController 간 데이터 전달방법
🤔 출품작 목록에서 출품작 상세 내용으로 이동할 때 해당 출품작에 대한 정보를 새로운 `ViewController`로 전달해야 했습니다. 이때 데이터를 전달할 방법에 대해 아래와 같은 방법들을 고려했습니다.
1. `init`으로 전달
    ```swift
    // ExpositionItemDetailViewController.swift (세번째 화면)
    init(expositionItemDetail: ExpositionItemEntity) {
        self.expositionItemDetail = expositionItemDetail
        super.init(nibName: nil, bundle: nil)
    }
    ```
    ```swift
    // ExpositionItemViewController.swift (두번째 화면)
    let detailViewController = ExpositionItemDetailViewController(expositionItemDetail: expositionItemEntity[indexPath.row])
    ```
    - 새로운 `ViewController`에서 데이터를 전달받을 수 있도록 `init`을 생성하여 전달받는 방법이 비교적 간단하다고 생각되었습니다.
2. 새로운 `ViewController` 객체에 `setter`를 만들거나 프로퍼티 자체에 접근해서 데이터 전달
    ```swift
    let detailViewController = ExpositionItemDetailViewController()
    detailViewController.expositionItemDetail = expositionItemEntity[indexPath.row]
    ```
    - 간단한 방법이지만 한 `ViewController`에서 다른 `ViewController`의 객체에 직접 접근하는 방식이 바람직한지에 대한 의구심이 들었습니다.
3. `delegate` 패턴으로 전달
    - `ItemDetailView`에서 이전 화면인 `ItemView`에 접근하여 데이터를 가져오는 것 자체는 가능하지만, 현재 `TableView` 내부에서 선택된 데이터가 무엇인지 특정하는 것이 불가능하다 판단하였습니다.

💡 결과적으로 1번(init으로 전달하는 방법)을 선택하였습니다.

### 🔹 errorLabel
- 처음에는 데이터를 받아오지 못한 경우 화면에 "데이터를 불러오지 못했습니다"라는 Label을 표시하도록 구현했습니다.
    ```swift
    private func configureErrorLabel() {
        let errorLabel = {
            let label = UILabel()
            label.translatesAutoresizingMaskIntoConstraints = false
            label.text ="데이터를 불러오지 못했습니다."
            return label
        }()
    }
    ```
    <img width="200" alt="스크린샷 2023-06-29 오후 5 49 46" src="https://github.com/yagom-academy/ios-exposition-universelle/assets/124643545/f8082a72-3d07-4b42-b088-ec6e26da48ce">
- 데이터의 일부 파일만 Decoding 되었을 때는 Decoding 된 파일은 보이는 게 더 적절하다고 판단하여 수정하였습니다
    - 실행화면 예시 : 

    | 정상 화면 | 박람회 정보 데이터 누락 | 포스터 누락 |
    | :--------: | :--------: | :--------: |
    | <img width="200" src="https://github.com/yagom-academy/ios-exposition-universelle/assets/124643545/b8c23ff5-d09c-4cc2-9a3a-c078f065e84f"> | <img width="200" src="https://github.com/yagom-academy/ios-exposition-universelle/assets/124643545/5c0d0bea-4397-445b-a7ce-de8a454a65c2"> | <img width="200" src="https://github.com/yagom-academy/ios-exposition-universelle/assets/124643545/dbedeb8b-65ed-4fa7-8946-bf5dff52daef"> |

<br>

## 🎙️ 조언을 얻고 싶은 점
### 🔸 요소의 재사용
태극기 버튼 부분을 구현하면서, `flag` 이미지가 두 번 반복되니 같은 `ImageView`를 반복해서 넣어줄 수도 있지 않을까 생각하였지만 불가능하다는 것을 확인했습니다. 화면을 그릴 때 동일 객체가 재사용되지 못하는 이유를 찾아보고 싶었는데, 정확한 원인을 찾기가 어려웠습니다. 이 부분에 대해서 조언을 구할 수 있을까요?

### 🔸 데이터 수정 범위
`JSON` 파일에 들어 있는 데이터 내용과 실행 예시 화면에 차이가 있는 것을 발견했습니다. 일부는 아래와 같이 해결을 하였는데, 마지막 항목인 문장 차이 부분은 `UI`단에서 고쳐야 하는 내용이 맞는지 판단이 되지 않아 의견을 여쭤보려고 남겨두었습니다.
- 방문객 수 : `UInt` 데이터에 `formatter`를 이용해서 천단위 구분기호를 추가하였습니다. 
- 박람회 `Title` 개행 : 데이터를 그대로 활용하면 아래와 같이 요구사항과 다른 방식으로 개행이 되어, 중간에 \n 이스케이프 시퀀스를 추가하였습니다.

    | 그대로 불러왔을 때 | 요구사항 |
    | :--------: | :--------: |
    | <img width="200" src="https://github.com/yagom-academy/ios-exposition-universelle/assets/124643545/9316842f-3dd5-484b-bf10-6c3c2026a90a"> | <img width="200" src="https://github.com/yagom-academy/ios-exposition-universelle/assets/124643545/3f41b79c-ba1e-49ba-9595-469d10b6cb0c"> |
    ```swift
    mainStackView.titleLabel.text = expositionEntity.title.replacingOccurrences(of: "(", with: "\n(")
    ```
- `description` 문장 형태 : `JSON 데이터`의 문장은 했다/했습니다  형태가 혼용되어 있는 것을 확인했습니다. 이럴 경우 문장의 어미까지 UI에서 수정을 하는 게 맞을까요? 이 부분은 단순히 보여지는 형태를 위한 포맷팅 처리가 아닌 데이터 내용의 영역이라고 생각되어 고민이 됩니다.

WhalesJin committed:
feat: NavigationController 생성

- 코드 베이스 UI 구현을 위하여 스토리보드 삭제 및 필요한 코드 구현

maxhyunm committed:
feat: setUpEntity 메서드 생성

- JSON 데이터를 변환한 expositionEntity 프로퍼티 생성
- 예외처리를 위한 isSetUpEntity 프로퍼티 생성

WhalesJin committed:
feat: ScrollView, StackView 생성

maxhyunm committed:
feat: configureStackView 메서드 생성

- titleLabel, posterImageView 구현
- isSetUpEntity가 false일 때를 위한 errorLabel 구현

WhalesJin committed:
feat: StackView에 descriptionLabel 추가

- descriptionLabel 추가에 따른 constraints 사항 추가

maxhyunm committed:
feat: configureDetailsStackView 메서드 생성, configureStackViewFromEntity 메서드 분리

WhalesJin committed:
feat: configureButtonStackView 메서드 구현

maxhyunm committed:
feat: ExpositionItemViewController 생성, tableView 기본 구현

- setUpEntity 메서드 구현하여 items 데이터 불러오기 완료

WhalesJin committed:
feat: ExpositionItemTableViewCell 클래스 구현

- name, image, shortDescription 프로퍼티 구현

maxhyunm committed:
feat: didTapButton 메서드 구현

- didTapButton 메서드로 changeViewButton을 눌렀을 때 화면 전환 구현
- buttonStackView 레이아웃 변경

WhalesJin committed:
refactor: ExpositionViewController의 UI 요소 레이아웃 수정

maxhyunm committed:
feat: ExpositionItemViewController 테이블뷰 상세 구현

- items JSON 데이터 변환 및 셀 내부에 삽입
- 테이블뷰 UI 레이아웃 수정

WhalesJin committed:
feat: ExpositionItemDetailViewController 생성

- 각 출품작 셀 터치 시 해당 작품의 DetailView로 전환

maxhyunm committed:
feat: ExpositionItemDetailViewController 구현

- 해당 작품의 imageName, description 기반으로 StackView 구성

WhalesJin committed:
feat: errorLabel 구현

- data 불러오기 실패시 띄울 label 구현

maxhyunm committed:
refactor: 매직리터럴 수정

- Namespace 생성

WhalesJin committed:
feat: DecodingManager 구조체 생성

- DecodingManager 메서드를 이용해서 각 뷰컨트롤러의 프로퍼티 구현

maxhyunm committed:
refactor: ChangeViewButtonStackView 분리

WhalesJin committed:
refactor: ExpositionInformationStackView 분리

maxhyunm committed:
refactor: MainStackView 분리

WhalesJin committed:
refactor: addSubView와 constraint 분리

maxhyunm committed:
refactor: 태극기 이미지 사이즈 조정

WhalesJin committed:
refactor: 파일 순서 및 네이밍 오타 등 수정

maxhyunm committed:
refactor: Namespace에서 errorMessage 삭제

WhalesJin committed:
docs: README.md 작성

stevenkim18 commented on Expo1900/Expo1900/Controller/ExpositionViewController.swift:
```diff
@@ -0,0 +1,76 @@
+import UIKit
+
+class ExpositionViewController: UIViewController {
+    private let expositionEntity = DecodingManager().decodeExpositionJSON()
```
프로터티 선언과 동시에 json 디코딩 된 값을 할당 하셨는데
이렇게 하신 이유가 있으실까요? 
다른 방법은 없을까요?

maxhyunm replied:
- 프로퍼티를 초기화하지 않으면 `init`을 생성하거나 옵셔널 처리를 해야 했는데, 아래와 같은 맥락에서 초기화시 데이터를 지정하는 것이 가장 나을 것 같다고 생각했습니다.
  - 해당 데이터는 처음 화면을 띄울 때부터 필요한 정보임
  - 생성자를 만들 경우 데이터 초기화 외에 해당 생성자의 쓰임새가 없음
  - 옵셔널 바인딩이 추가되면 뷰컨트롤러가 더 복잡해질 것이라고 판단
- 하지만 리팩토링을 진행하다 보니 `ExpositionViewController`에서는 특정 메서드 내에서만 해당 엔티티 정보가 필요한 것으로 판단되어, `expositionEntity`를 전역변수로 두지 않고 파라미터를 통해 전달하는 방식으로 수정하였습니다.
- `DecodingManager`를 리팩토링하면서 do-catch문이 필요해짐에 따라 `ExpositionItemViewController`에서는 엔터티를 위한 변수를 빈 배열로 초기화한 뒤 `viewDidLoad`에서 받아오는 방식으로 수정하였습니다.
- 관련 커밋 : 76b3f8e7dfbfc2d516956f0c27eac2d87b344628

stevenkim18 commented on Expo1900/Expo1900/Controller/ExpositionViewController.swift:
```diff
@@ -0,0 +1,76 @@
+import UIKit
+
+class ExpositionViewController: UIViewController {
+    private let expositionEntity = DecodingManager().decodeExpositionJSON()
```
지금 케이스에서는 처음 진입하는 화면이라서 그럴 경우가 없겠지만
해당 화면이 생성 되었다 안되었다 하면 `DecodingManager()` 인스턴스도 계속 생성 됩니다.

디코딩을 하는 것은 앱의 전반적으로 사용될 것 같은데 인스턴스를 한번 만 생성하고 앱 내에서 쭉 사용하는 방법은 없을까요?

WhalesJin replied:
해당 부분은 싱글톤 방식으로 수정하여 인스턴스를 한 번만 생성하도록 수정하였습니다.
관련 커밋 : 76b3f8e7dfbfc2d516956f0c27eac2d87b344628

stevenkim18 commented on Expo1900/Expo1900/Controller/ExpositionViewController.swift:
```diff
@@ -0,0 +1,76 @@
+class ExpositionViewController: UIViewController {
+    private let expositionEntity = DecodingManager().decodeExpositionJSON()
+    
+    private let mainScrollView = {
+        let scrollView = UIScrollView()
+        scrollView.translatesAutoresizingMaskIntoConstraints = false
+        
+        return scrollView
+    }()
```
제 취향이기는 한데.. 
view들을 정의 할 때는 타입을 명시하는 편입니다. 
그래야 좀 덜 헷갈려요 ㅎㅎ

```swift
private let mainScrollView: UIScrollView = {
        let scrollView = UIScrollView()
        scrollView.translatesAutoresizingMaskIntoConstraints = false

        return scrollView
    }()
```

maxhyunm replied:
주신 조언 참고해서 클로저로 구현을 하는 경우에는 타입 명시를 다 해두고
인스턴스로 바로 생성하는 경우에는 타입이 같이 보여서 괜찮다고 생각이 들어서 따로 타입을 명시하지 않았습니다 😃

stevenkim18 commented on Expo1900/Expo1900/Controller/ExpositionViewController.swift:
```diff
@@ -0,0 +1,76 @@
+            mainScrollView.leftAnchor.constraint(equalTo: view.leftAnchor),
+            mainScrollView.rightAnchor.constraint(equalTo: view.rightAnchor),
+            
+            mainStackView.topAnchor.constraint(equalTo: mainScrollView.topAnchor, constant: 20),
+            mainStackView.bottomAnchor.constraint(equalTo: mainScrollView.bottomAnchor, constant: -20),
+            mainStackView.centerXAnchor.constraint(equalTo: mainScrollView.centerXAnchor),
+            mainStackView.widthAnchor.constraint(equalTo: mainScrollView.widthAnchor, constant: -40),
```
문자열 상수를 namespace로 분리해주신 것 처럼 
숫자 상수도 가능하시면 분리하시면 좋습니다 ㅎㅎ

WhalesJin replied:
추후의 수정을 용이하게 하기 위하여 주신 리뷰대로 숫자용 `namespace`를 분리하였습니다!!
관련 커밋 : 207077d27a943a641c10c03bb4c24c02eee25dc1

stevenkim18 commented on Expo1900/Expo1900/Controller/Namespace/IdentifierNamespace.swift:
```diff
@@ -0,0 +1,10 @@
+//
+
+enum IdentifierNamespace {
+    static let cell = "cell"
```
어떤 cell을 말하시는 걸까요??

maxhyunm replied:
두번째 화면의 `itemTableView`의 `cell`이어서 `itemTableCell`로 네이밍 수정하였습니다!!😁

stevenkim18 commented on Expo1900/Expo1900/Controller/ExpositionItemViewController.swift:
```diff
@@ -0,0 +1,73 @@
+            return ExpositionItemTableViewCell()
+        }
+        
+        cell.image.image = UIImage(named: expositionItemEntity[indexPath.row].imageName)
+        cell.name.text = expositionItemEntity[indexPath.row].name
+        cell.shortDescription.text = expositionItemEntity[indexPath.row].shortDescription
+        
+        cell.accessoryType = .disclosureIndicator
```
cell의 대한 값들은 cell 안에서 처리하는 것이 좋아요.

필요한 데이터만 cell의 함수로 넘겨주고
cell안에 있는 뷰에 대한 요소들은 
cell의 함수 안에서 처리하는 것이 
책임과 역할의 분리 측면에 더 좋다고 생각합니다.

WhalesJin replied:
`inputDataToCell` 메서드를 생성하여 그 안에 요소들을 구현하였습니다!! 감사합니다 😄👍🏻
관련 커밋 : b1dca5893a078a53a41e4171a14184fa8f96739c

stevenkim18 commented on Expo1900/Expo1900/Controller/ExpositionItemViewController.swift:
```diff
@@ -0,0 +1,73 @@
+        return cell
+    }
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let detailViewController = ExpositionItemDetailViewController(
+            expositionItemDetail: expositionItemEntity[indexPath.row]
+        )
+        
+        navigationController?.pushViewController(detailViewController, animated: true)
+    }
```
뷰컨에서 다른 뷰컨으로 데이터를 넘겨주실 때 생성자를 통해서 넘겨주셨네요!
잘하셨습니다!! 

혹시 다른 방법은 없었을까요? 혹시 있다면 생성자로 넘겨주는 것과 어떤 차이점이 있을까요?

maxhyunm replied:
이 내용은 하단에 따로 답변드리겠습니다!

stevenkim18 commented on Expo1900/Expo1900/Controller/ExpositionItemDetailViewController.swift:
```diff
@@ -0,0 +1,85 @@
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
```
`required init?`은 왜 필요할까요?

WhalesJin replied:
`NSCoder`를 전달받는 `required init`은 스토리보드 기반 뷰에서 필요한 생성자입니다. 실패 가능한 이니셜라이저라 `init` 뒤에 `?` 가 붙었습니다. `NSCoder`에서 스토리보드의 요소들을 코드로 변환해주는 역할을 하기에, 모든 뷰에 생성자를 추가할 경우 이 생성자를 필수적으로 구현해주어야 합니다.
보통 `required init`이 들어가는 이유는 해당 클래스를 상속받은 자식 클래스에서도 반드시 구현해야 하는 내용이나 초기화해야하는 프로퍼티가 있을 경우 사용하게 됩니다.
하지만 코드베이스로 뷰를 구현할 경우, `NSCoder`가 들어간 생성자를 거칠 필요가 없어집니다. 때문에 이 경우에는 구현부에 `fatalError` 호출부가 들어갑니다. 해당 메서드를 거치는 순간 앱의 모든 흐름이 종료되기 때문에, `coder` 프로퍼티를 초기화하지 않고도 정상적으로 작동할 수 있습니다.

stevenkim18 commented on Expo1900/Expo1900/Model/DecodingManager.swift:
```diff
@@ -0,0 +1,32 @@
+
+import UIKit
+
+struct DecodingManager {
+    let decoder: JSONDecoder = JSONDecoder()
+    
+    func decodeExpositionJSON() -> ExpositionEntity {
+        guard let exposition: NSDataAsset = NSDataAsset(name: AssetNamespace.expositionUniverselle),
+              let decodedExposition: ExpositionEntity = try? decoder.decode(ExpositionEntity.self, from: exposition.data)
+        else {
+            return ExpositionEntity.init(title: "", visitors: 0, location: "", duration: "", description: "")
+        }
+        
+        return decodedExposition
+    }
+    
+    func decodeExpositionItemsJSON() -> [ExpositionItemEntity] {
+        guard let expositionItems: NSDataAsset = NSDataAsset(name: AssetNamespace.items),
+              let decodedExpositionItems: [ExpositionItemEntity] = try? decoder.decode([ExpositionItemEntity].self, from: expositionItems.data)
+        else {
+            return [ExpositionItemEntity.init(name: "", imageName: "", shortDescription: "", description: "")]
+        }
+        
+        return decodedExpositionItems
+    }
+}
```
`DecodingManager`로 따로 분리하신 것은 매우 잘하셨습니다!👍

이 타입을 조금 더 간결하고 유연하게 만들 수 있을 것 같아요.
지금 같은 경우는 새로 디코딩할 타입이 추가되면 함수를 계속 만들어줘야하는데
타입에 유연하게 함수 한개로 처리할 수 있는 방법은 없을까요??

maxhyunm replied:
두 가지 메서드의 반환타입이 달라 따로 구현하였는데 주신 리뷰를 보고 방법이 있을거라 생각되어 고민해보다가 `Generic Type`을 적용하였습니다. 
- `Generic Type`
    - 여러가지 타입에 작동하는 코드를 만들기 위한 방법
    - '종류는 알 수 없지만 어떠한 타입이 여기에 쓰일 것이다' 라는 표현
    - 어떤 타입에도 유연하게 대응하는 코드를 작성할 수 있음
    - 재사용하기 쉽고 코드의 중복을 줄일 수 있어 깔끔하고 추상적인 표현이 가능

저희가 구현한 `DecodingManager`의 경우 `JSON decode`가 가능한 형태의 타입이 와야 하기 때문에, `Decodable`을 채택한 타입만 올 수 있도록 `<T: Decodable>` 형식으로 지정하였습니다.

- 관련 커밋 : 76b3f8e7dfbfc2d516956f0c27eac2d87b344628

stevenkim18 commented on Expo1900/Expo1900.xcodeproj/project.pbxproj:
```diff
@@ -81,12 +102,47 @@
 			path = Expo1900ModelTests;
 			sourceTree = "<group>";
 		};
+		3CBDC7D52A4D0E5F0089F7A2 /* Controller */ = {
+			isa = PBXGroup;
+			children = (
+				BAFD9F502A4D83A000DC9186 /* Namespace */,
+				C79FF4B42589F401005FB0FD /* AppDelegate.swift */,
+				C79FF4B62589F401005FB0FD /* SceneDelegate.swift */,
+				C79FF4B82589F401005FB0FD /* ExpositionViewController.swift */,
+				BAFD9F4E2A4D0A7500DC9186 /* ExpositionItemViewController.swift */,
+				3CBDC8242A4D6D560089F7A2 /* ExpositionItemDetailViewController.swift */,
+			);
+			path = Controller;
+			sourceTree = "<group>";
+		};
+		3CBDC7D62A4D0F380089F7A2 /* View */ = {
```
`AppDelegate`와 `SceneDelegate`는 Controller라고 생각하셔서 해당 폴더에 넣으신건가요?

WhalesJin replied:
`AppDelegate`와 `SceneDelegate`는 앱과 윈도우를 관리하기 때문에 
`MVC` 중 그 어떤 것에도 속하지 않는다고 생각해서 `MVC` 바깥 파일로 분류하였습니다.
관련 커밋 : 8b9dfa98391c5bab01ea30dd1034aec9f0321c27

maxhyunm replied:
`AppDelegate`와 `SceneDelegate`는 앱과 윈도우를 관리하기 때문에 
`MVC` 중 그 어떤 것에도 속하지 않는다고 생각해서 `MVC` 바깥 파일로 분류하였습니다.
- 관련 커밋 : 8b9dfa98391c5bab01ea30dd1034aec9f0321c27

stevenkim18 reviewed (commented):
맥스, 웨일 스탭2 하시느라 고생 많으셨습니다!

스탭1 보다 봐야할 것들이 많아서 시간이 조금 더 걸린 것 같아요.

일단 한 번 쭉 보고 생각 것들을 적어봤습니다.

조금 더 자세히 보고 추가로 리뷰 할 것들이 있으면 아래 추가 댓글로 달아둘게요!

---

# 고민했던 점에 대한 답변

> **UI 구현 - Storyboard vs Code**
🤔 스토리보드는 이미 많은 부분을 자동으로 해주고 있어서 참 편리한 친구라고 생각하지만 스토리보드로만 할 수 있는 부분에 대해서는 한계가 있기에 코드로 구현하는 부분들을 더 활용하고 싶어서 공부를 하고 적용해 보았습니다.
> 
> - 스토리보드와 코드를 비교했을 때 코드의 장점
>     - 나중에 소스로 볼 때
>         - 스토리보드 : XML이라 보기 힘들다
>         - 코드 : 코드로 적은 거라 더 깔끔하고 이해하기 좋다
>     - 빌드
>         - 스토리보드 -> 코드 -> 시뮬레이터
>         - 코드 -> 시뮬레이터 (빌드 속도가 좀 더 빠르다)
> - 코드베이스로 구현한 이유
>     - 커스텀으로 `TableViewCell`과 `View`들을 만들어 적용해보기 위해(`View` 파일 분리)
>     - 기존에 스토리보드 형태로만 작업을 해 보았기 때문에 새로운 학습을 위해 시도

일단 장단점을 비교하신 것은 매우 잘하셨습니다 👍

소스로 볼때 장단점을 비교하셨는데 실제로 스토리보드를 소스로 보는 일은 잘 없습니다 ㅎㅎ

소스 보다는 오히려 기획된 화면이 히스토리가 남지 않는 경우 

스토리보드가 있으면 오히려 파악하기 더 유리합니다. 코드로 보려면 빌드를 해야되지만 스토리보드는 빌드 없이 볼 수 있습니다.

그리고 “빌드 속도가 더 빠르다” 라고 하셨는데 이런 기술에 대한 것들을 비교할 때는 실제 수치를 가져오는 것도 좋습니다. 몇초가 더 빨랐는지, 규모에 따라서 어떻게 속도가 차이났는지에 대한 정확한 근거를 가져오시면 더 좋았을 것 같습니다.

실제로 지금 하시는 프로젝트에서는 스토리보드나 코드나 빌드 속도의 큰 차이는 나지 않을 껍니다. 

그리고 지금 m1 맥이 나와서 스토리보드와 코드의 빌드 속도는 점점 줄어 들고 있습니다.

스토리보드가 무거워지면 개별 뷰컨이나 뷰들을 xib 파일(스토리보드와 비슷) 분리할 수 있습니다.

> 레이아웃 관리
> 
> - **`StackView` 내부 요소가 화면 영역보다 커지는 경우**
>     
>     🚨 `StackView`의 `leftAnchor`, `rightAnchor`를 맞추었음에도 불구하고 `StackView` 내부 요소가 커지면 상위 요소들이 함께 늘어나는 것을 확인했습니다.
>     
>     💡 `StackView`의 `width`를 지정하여 해결하였습니다.
>     
>     `mainStackView.widthAnchor.constraint(equalTo: mainScrollView.widthAnchor, constant: -40)`
>     
> - **`StackView` 내부 요소가 잘리거나 사라지는 경우**
>     
>     🚨 `horizontal`로 지정된 `StackView`에서 하나의 요소가 커지면 다른 요소가 잘리거나 사라지는 것을 확인했습니다.
>     
>     💡 `setContentCompressionResistancePriority` 설정으로 텍스트버튼의 resistance 우선순위를 높였습니다.
>     
>     💡 태극기의 높이를 버튼 높이와 맞추고, 태극기의 가로 길이는 높이와 비율을 계산하여 맞추어 해결하였습니다.
>     

이 부분은 코드와 함께 before after 화면이 있었으면 제가 빠르게 이해하기 좋았을 것 같아요 ㅎㅎ

maxhyunm committed:
refactor: closure로 구현된 UI 구성요소의 타입 명시

stevenkim18 commented:
맥스, 웨일!! 추가적으로 리뷰 남깁니다!! 확인부탁드려요🙏

## 고민 했던 점에 대한 답변

> ViewController 기능 분리
> 
> 
> 🤔 요소를 구현할 때마다 관련 인스턴스를 생성하여 추가하고 제약조건을 넣다 보니 ViewController가 너무 무거워지는 것을 느꼈습니다. 때문에 아래 방법들로 ViewController의 정리 및 기능 분리를 시도했습니다.
> 
> - View 파일 분리
>     - 내부 요소로 추가되는 중첩 `StackView`를 커스텀 `StackView`로 분리했습니다.
>         - `MainStackView`
>         - `ExpositionInformationStackView`
>         - `ChangeViewButtonStackView`

뷰컨이 무거워지는 것을 인지하시고 분리하려는 시도는 매우 좋습니다👍

다만 개인적으로 스택뷰를 너무 많이 사용하신 것이 아닌가 싶습니다. 그리고 오히려 커스텀 스택뷰로 분리하셔서 코드를 한 눈에 파악하기 어려웠습니다 😭

세가지 질문을 드릴게요!

1. 커스텀 뷰는 언제 만드는 것이 좋을까요?
2. 스택뷰는 언제 사용하는 것이 좋을까요?

> 🔹 ViewController 간 데이터 전달방법
> 
> 
> 🤔 출품작 목록에서 출품작 상세 내용으로 이동할 때 해당 출품작에 대한 정보를 새로운 `ViewController`로 전달해야 했습니다. 이때 데이터를 전달할 방법에 대해 아래와 같은 방법들을 고려했습니다.
> 
> 1. `init`으로 전달
>     
>     `// ExpositionItemDetailViewController.swift (세번째 화면)
>     init(expositionItemDetail: ExpositionItemEntity) {
>         self.expositionItemDetail = expositionItemDetail
>         super.init(nibName: nil, bundle: nil)
>     }`
>     
>     `// ExpositionItemViewController.swift (두번째 화면)
>     let detailViewController = ExpositionItemDetailViewController(expositionItemDetail: expositionItemEntity[indexPath.row])`
>     
>     - 새로운 `ViewController`에서 데이터를 전달받을 수 있도록 `init`을 생성하여 전달받는 방법이 비교적 간단하다고 생각되었습니다.
> 2. 새로운 `ViewController` 객체에 `setter`를 만들거나 프로퍼티 자체에 접근해서 데이터 전달
>     
>     `let detailViewController = ExpositionItemDetailViewController()
>     detailViewController.expositionItemDetail = expositionItemEntity[indexPath.row]`
>     
>     - 간단한 방법이지만 한 `ViewController`에서 다른 `ViewController`의 객체에 직접 접근하는 방식이 바람직한지에 대한 의구심이 들었습니다.
> 3. `delegate` 패턴으로 전달
>     - `ItemDetailView`에서 이전 화면인 `ItemView`에 접근하여 데이터를 가져오는 것 자체는 가능하지만, 현재 `TableView` 내부에서 선택된 데이터가 무엇인지 특정하는 것이 불가능하다 판단하였습니다.
> 
> 💡 결과적으로 1번(init으로 전달하는 방법)을 선택하였습니다.
> 

3가지 방법 중에 1번 방법이 간단해서 선택하셨다는 근거가 조금 부족한 것 같습니다. 

3가지 방법에 대해서 차이점을 조금 더 공부해보시면 좋을 것 같습니다

(3번 방법에서 "tableview 내부에서 선택 된 데이터가 무엇인지 특정하는 것” 이 어떤 말인지 잘 이해가 가지 않습니다)

> errorLabel
> 
> - 처음에는 데이터를 받아오지 못한 경우 화면에 "데이터를 불러오지 못했습니다"라는 Label을 표시하도록 구현했습니다.
> 
> 데이터의 일부 파일만 Decoding 되었을 때는 Decoding 된 파일은 보이는 게 더 적절하다고 판단하여 수정하였습니다
> 

데이터를 불러오지 못한 경우 예외처리 하신 것은 매우 좋습니다!! 사용자를 고려하셨군요 👍

하나 궁금한 점은 데이터가 일부만 디코딩 되는 경우도 일부의 데이터를 보여주는 것이 좋겠다고 판단한 이유가 궁금합니다.

저라면 그냥 데이터를 못 불러왔다고 할 것 같아서요 ㅎㅎ

## 조언을 얻고 싶은 점에 대한 답변

> 🔸 요소의 재사용
> 
> 
> 태극기 버튼 부분을 구현하면서, `flag` 이미지가 두 번 반복되니 같은 `ImageView`를 반복해서 넣어줄 수도 있지 않을까 생각하였지만 불가능하다는 것을 확인했습니다. 화면을 그릴 때 동일 객체가 재사용되지 못하는 이유를 찾아보고 싶었는데, 정확한 원인을 찾기가 어려웠습니다. 이 부분에 대해서 조언을 구할 수 있을까요?
> 

객체와 인스턴스의 개념에 대해서 헷갈리신 것 같습니다.

한 개의 인스턴스를 재활용할 수는 없습니다.

태극기를 보여주는 `imageView`는 별 다른 기능을 하지 않기때문에 함수 한 개를 만들어서 2번 호출하는 방법이 있을 것 같습니다.

```swift
func flagImageView() -> UIImageView {
	let imageView = UIImageView()
        imageView.image = UIImage(named: AssetNamespace.flag)
        
   return imageView
}
```

> 🔸 데이터 수정 범위
> 
> 
> `JSON` 파일에 들어 있는 데이터 내용과 실행 예시 화면에 차이가 있는 것을 발견했습니다. 일부는 아래와 같이 해결을 하였는데, 마지막 항목인 문장 차이 부분은 `UI`단에서 고쳐야 하는 내용이 맞는지 판단이 되지 않아 의견을 여쭤보려고 남겨두었습니다.
> 
> - 방문객 수 : `UInt` 데이터에 `formatter`를 이용해서 천단위 구분기호를 추가하였습니다.
> - 박람회 `Title` 개행 : 데이터를 그대로 활용하면 아래와 같이 요구사항과 다른 방식으로 개행이 되어, 중간에 \n 이스케이프 시퀀스를 추가하였습니다.
>     
>     `mainStackView.titleLabel.text = expositionEntity.title.replacingOccurrences(of: "(", with: "\n(")`
>     
> - `description` 문장 형태 : `JSON 데이터`의 문장은 했다/했습니다 형태가 혼용되어 있는 것을 확인했습니다. 이럴 경우 문장의 어미까지 UI에서 수정을 하는 게 맞을까요? 이 부분은 단순히 보여지는 형태를 위한 포맷팅 처리가 아닌 데이터 내용의 영역이라고 생각되어 고민이 됩니다.

뷰컨과 entity 사이에 모델을 하나 더 두면 어떨까요?

뷰컨는 해당 모델이 주는 값만 화면에 보여주고 

그 모델을 entity를 들고 있고 위에서 말씀하신 로직들을 처리하는 겁니다.

(마지막 `description` 에 했다/했습니다 는 데이터의 영역인 것 같습니다. ㅎㅎ 지금은 크게 신경쓰지 않으셔도 됩니다)

WhalesJin committed:
refactor: 매직넘버 수정

- Namespace 생성(Spacing, Constraints, Multiplier)

maxhyunm committed:
refactor: inputDataToCell 메서드 생성

WhalesJin committed:
refactor: 데이터 전처리를 위한 String 확장

maxhyunm committed:
refactor: MainStackView 타입 삭제

WhalesJin committed:
refactor: DecodingManager 타입 수정

- DecodingManager 싱글톤 변경
- DecodingError 타입 생성 -> do-catch문
- decodeJSON 메서드를 제네릭 타입으로 수정

maxhyunm committed:
refactor: ErrorLabel 타입 생성

- JSON 디코딩 실패시 에러메시지 표시

WhalesJin committed:
refactor: 반복되는 FlagImageView를 메서드로 구현

maxhyunm committed:
chore: AppDelegate, SceneDelegate 파일 위치 변경

maxhyunm committed:
refactor: DecodingManager 누락된 init 추가

WhalesJin commented:
### 질문 주신 부분에 대한 답변

### 🔹 ViewController 기능 분리

>뷰컨이 무거워지는 것을 인지하시고 분리하려는 시도는 매우 좋습니다👍
다만 개인적으로 스택뷰를 너무 많이 사용하신 것이 아닌가 싶습니다. 그리고 오히려 커스텀 스택뷰로 분리하셔서 코드를 한 눈에 파악하기 어려웠습니다 😭
세가지 질문을 드릴게요!
>>1. 커스텀 뷰는 언제 만드는 것이 좋을까요?
>>2. 스택뷰는 언제 사용하는 것이 좋을까요?
>>3. 스택뷰를 사용하지 않고 구현하는 방법은?

> Q1. 커스텀 뷰는 언제 만드는 것이 좋을까요?
- 비슷한 형태의 뷰가 여러 번 활용될 때 커스텀 뷰를 사용하면 좋습니다.(재사용성)
- 한 뷰 안에 구현될 내용이 많아서 분리가 필요할 때에도 사용할 수 있습니다.

> Q2. 스택뷰는 언제 사용하는 것이 좋을까요?
- 두 개 이상의 요소들이 수평 혹은 수직으로 일렬 배치되며 그 간격이 일정할 때 사용하기 좋습니다. (필요에 따라 스페이싱 커스텀 가능)
- 오토 레이아웃을 활용하여 장치의 방향, 화면 크기 등의 변화에 동적으로 적응할 수 있는 사용자 인터페이스를 만들 때 스택뷰를 사용합니다.
- 참고링크 : [UIStackView](https://developer.apple.com/documentation/uikit/uistackview)

> Q3. 스택뷰를 사용하지 않고 구현하는 방법은?
- 각 요소의 레이아웃 제약조건과 priority를 설정합니다.


답변을 준비하며 고민을 한 끝에, `mainStackView`는 커스텀타입으로 구현할 필요는 없다는 생각이 들었습니다. 다만 `ScrollView` 안에서 전체 요소를 모아주는 뷰가 필요하다는 생각에 `mainStackView` 자체를 제거하지는 않고, 뷰컨트롤러 내부에서 구현하는 방식으로 변경하였습니다.
```swift
private let mainStackView: UIStackView = {
        let stackView = UIStackView()
        stackView.translatesAutoresizingMaskIntoConstraints = false
        stackView.axis = .vertical
        stackView.alignment = .center
        stackView.spacing = SpacingNamespace.mainStackView
        
        return stackView
    }()
```
또한 뷰컨트롤러의 가독성을 고려하여 StackView 안에 요소들이 추가되는 부분과 제약조건이 추가되는 부분을 `ExpositionViewController`의 `extension`으로 분리하였습니다.

```swift
extension ExpositionViewController {
    private func configureMainView() {
        ...
    }
    private func configureLabelText(_ expositionEntity: ExpositionEntity) {
        ...
    }
    private func addConstraints() {
        ...
    }                          
```
`mainStackView` 내부에 요소로 포함되는 `ExpositionInformationStackView`, `ChangeViewButtonStackView`의 경우 가로 방향으로 두 가지 이상의 요소가 배치되면서 동시에 중앙정렬이 되어야 하는 점을 감안해 `StackView` 형식을 그대로 살려두었습니다. 또한 아래와 같은 점을 감안하여 두 가지 모두 커스텀 타입으로 진행하였습니다.
    - `ExpositionInformationStackView`는 같은 형태가 세 번 반복되어 재사용이 되고 있음
    - `ChangeViewButtonStackView`는 다른 내용과 구분되는 버튼을 구현하는 내용으로, 별도의 타입으로 나뉘어도 괜찮다고 판단
- 관련 커밋 : 7f7a9c8cd8c97a290d029fc3950c47d94678736e

### 🔹 ViewController 간 데이터 전달방법

> 3가지 방법 중에 1번 방법이 간단해서 선택하셨다는 근거가 조금 부족한 것 같습니다.
3가지 방법에 대해서 차이점을 조금 더 공부해보시면 좋을 것 같습니다
(3번 방법에서 "tableView 내부에서 선택 된 데이터가 무엇인지 특정하는 것"이 어떤 말인지 잘 이해가 가지 않습니다)
- 생성자를 활용한 가장 큰 이유 중 하나는 생성자를 통해 데이터 전달시 해당 뷰컨트롤러를 초기화할 때 데이터 내용 또한 초기화가 가능하기 때문입니다. 여기서 전달받는 데이터는 세 번째 화면을 그리는데 있어 필수적인 요소이기 때문에, 옵셔널 처리를 해야 하는 다른 방식보다는 생성자를 통한 전달이 효율적이라고 판단했습니다.
- `tableView 내부에서 선택된 데이터가 무엇인지 특정하는 것`의 의미
    - `tableView`에서 특정 셀을 선택시 해당 데이터를 세 번째 뷰로 가져와야 하는 상황이기 때문에, 만약 세 번째 뷰컨트롤러에서 델리게이트를 통해 두 번째 뷰컨트롤러에 접근해 데이터를 받아오게 된다면 어떤 셀에 포함된 데이터를 가져와야 하는지를 파악하기 어렵다는 의미로 작성하였습니다.



### 🔹 errorLabel
>데이터를 불러오지 못한 경우 예외처리 하신 것은 매우 좋습니다!! 사용자를 고려하셨군요 👍
하나 궁금한 점은 데이터가 일부만 디코딩 되는 경우도 일부의 데이터를 보여주는 것이 좋겠다고 판단한 이유가 궁금합니다.
저라면 그냥 데이터를 못 불러왔다고 할 것 같아서요 ㅎㅎ

예를 들어 다른 정보는 제대로 들어오는데 포스터만 잘못 들어온다면 나머지 정보라도 보여주는 게 어떤가 하는 판단에 일부 데이터라도 보여지도록 코드를 구현하였었습니다.
하지만 이번 리팩토링 과정에서 `DecodingManager`에 제네릭 타입을 추가하면서 에러처리가 필요하게 되었고 `DecodingError` 타입은 화면 구성 이전 단계에 발생하는 에러이기 때문에 빈 화면에 `ErrorLabel`만 띄우도록 코드를 수정하였습니다.

```swift
enum DecodingError: Error {
    case failedDecoding
    case unknown
    
    var message: String {
        switch self {
        case .failedDecoding:
            return "데이터를 불러오지 못했습니다."
        case .unknown:
            return "알 수 없는 오류입니다."
        }
    }
}
```
```swift
do { ... 
} catch DecodingError.failedDecoding {
    ErrorLabel(DecodingError.failedDecoding.message, frame: view.frame).configureUI(view)
} catch {
    ErrorLabel(DecodingError.unknown.message, frame: view.frame).configureUI(view)
}
```

<img src = "https://hackmd.io/_uploads/Bku60SZth.png" width = "200"> <img src = "https://hackmd.io/_uploads/HJOpRHZKh.png" width = "203">

### 🔸 요소의 재사용

> 태극기를 보여주는 imageView는 별 다른 기능을 하지 않기때문에 함수 한 개를 만들어서 2번 호출하는 방법이 있을 것 같습니다.

정말정말 좋은 방법인 것 같아서 참고하여 아래와 같이 수정하였습니다 👍🏻👍🏻
```swift
class ChangeViewButtonStackView: UIStackView {
    private func configureUI() {
                        ...
        let leftFlagView: UIImageView = createFlagImageView()
        let rightFlagView: UIImageView = createFlagImageView()
        
        self.addArrangedSubview(leftFlagView)
        self.addArrangedSubview(rightFlagView)
                        ...
    }
    
    func createFlagImageView() -> UIImageView {
                        ...
    }
}
```


### 🔸 데이터 수정 범위

> - 방문객 수 : `UInt` 데이터에 `formatter`를 이용해서 천단위 구분기호를 추가하였습니다. 
> - 박람회 `Title` 개행 : 데이터를 그대로 활용하면 아래와 같이 요구사항과 다른 방식으로 개행이 되어, 중간에 \n 이스케이프 시퀀스를 추가하였습니다.
>
> 뷰컨과 entity 사이에 모델을 하나 더 두면 어떨까요?
> 뷰컨은 해당 모델이 주는 값만 화면에 보여주고
> 그 모델을 entity를 들고 있고 위에서 말씀하신 로직들을 처리하는 겁니다.

주신 리뷰 참고하여 `String` 타입을 확장하여 진행하였습니다!!
관련 커밋 : 348545ff7d2bcac0003f6eb96aad9915cfdfc438

WhalesJin committed:
refactor: 추가 은닉화 및 DecodingManager 타입 변환

stevenkim18 commented:
제가 요청해 주신 부분들 잘 수정하셨습니다.
크게 이상한 것이 없어서 머지 하겠습니다.

추가로 코멘트 남긴 것은 스탭3에서 반영해주세요!

stevenkim18 commented:
> tableView 내부에서 선택된 데이터가 무엇인지 특정하는 것의 의미
tableView에서 특정 셀을 선택시 해당 데이터를 세 번째 뷰로 가져와야 하는 상황이기 때문에, 만약 세 번째 뷰컨트롤러에서 델리게이트를 통해 두 번째 뷰컨트롤러에 접근해 데이터를 받아오게 된다면 어떤 셀에 포함된 데이터를 가져와야 하는지를 파악하기 어렵다는 의미로 작성하였습니다.

`UITableView` 함수나 변수 중에 현재 선택된 cell의 `indexPath` 을 알 수 있는 것이 있을거에요 ㅎㅎ 아예 불가능한 방법은 아닐꺼에요 ㅎㅎ

위에서 말씀하신 것 처럼 지금 상황에서는 생성자를 통해서 데이터를 넘기는게 저도 좋아 보이네요 ㅎㅎ

나중에 시간이 되시면 `생성자 주입`이라는 키워드도 학습해보세요 🧐
